src/Microsoft.ML.Dnn/DnnRetrainTransform.cs

```diff
@@ -382,10 +383,10 @@ private void TrainCore(DnnRetrainEstimator.Options options, IDataView input, IDa
                 runner.AddInput(inputs[i], srcTensorGetters[i].GetBufferedBatchTensor());
 
             Tensor[] tensor = runner.Run();
-            var buffer = tensor[0].Data();
-            loss = tensor.Length > 0 && tensor[0] != IntPtr.Zero ? (float)tensor[0].Data<float>()[0] : 0.0f;
-            metric = tensor.Length > 1 && tensor[1] != IntPtr.Zero ? (float)tensor[1].Data<float>()[0] : 0.0f;
-            var b = tensor.Length > 2 && tensor[2] != IntPtr.Zero ? (float[])tensor[2].Data<float>() : null;
+            //var buffer = tensor[0].Data();
+            loss = tensor.Length > 0 && tensor[0] != IntPtr.Zero ? (float)tensor[0].ToArray<float>()[0] : 0.0f;
```

Comment:
I haven't looked at the implementation of `ToArray`, but I assume it copies the data out into a new array instance. That isn't great for performance reasons (creating a new array instance, copying the values over, and then GC'ing the instance).

/cc @codemzs

Reply:
The API change is from Data() to ToArray(). The Data<T>() already did a memcopy, so the performance would be unchanged with ToArray()

---
In reply to: [323474780](https://github.com/dotnet/machinelearning/pull/4205#discussion_r323474780) [](ancestors = 323474780)